File: MSVReverseSingleLinkedList/MSVReverseSingleLinkedList/RLLinkedList.m
//
//  RLLinkedList.m
//  ReverseList
//
//  Created by Serge Moskalenko on 09.06.16.
//  Skype:camopu-ympo
//  mob: +380677629137
//  Copyright (c) 2016 Serge Moskalenko. All rights reserved.
//

#import "RLLinkedList.h"

@implementation RLLinkedList


- (void)addNode:(RLNode* )node
{
    if (self.tail == nil)
        self.head = node;
    else
        self.tail.next = node;

    self.tail = node;
}

- (void)reverse
{
    RLNode* newTail = self.head;
    self.head = [self reverseNode:self.head];
    self.tail = newTail;
}

-(RLNode*)reverseNode:(RLNode*) root
{
    RLNode* newRoot = nil;
    while (root) {
        RLNode* next = root.next;
        root.next = newRoot;
        newRoot = root;
        root = next;
    }
    return newRoot;
}

- (NSString *)description
{
    NSMutableString* result = [NSMutableString string];
    RLNode* node = self.head;
    while (1)
    {
        [result appendFormat:@"%ld ", (long)node.data];
        node = node.next;
        if (node == nil)
            break;
    }
    
    return result;
}

@end

File: MSVReverseSingleLinkedList-swift/MSVReverseSingleLinkedList-swift/RLLinkedList.swift
//
//  RLLinkedList.h
//  ReverseList
//
//  Created by Serge Moskalenko on 09.06.16.
//  Skype:camopu-ympo
//  mob: +380677629137
//  Copyright (c) 2016 Serge Moskalenko. All rights reserved.
//

import Foundation
class RLLinkedList: NSObject {
    var head: RLNode? = nil
    weak var tail: RLNode?
    
    func addNode(node: RLNode) {
        if self.tail == nil {
            self.head = node
        }
        else {
            self.tail!.next = node
        }
        self.tail = node
    }
    
    func reverse() {
        let newTail: RLNode = self.head!
        self.head = self.reverseNode(self.head)
        self.tail = newTail
    }
    
    func customDescription() -> String {
        var result: String = ""
        var node: RLNode? = self.head
        while true {
            result += "\(node!.data) "
            node = node!.next
            if node == nil {
                break
            }
        }
        return result
    }
    
    override var description: String {
        return customDescription()
    }
    
    func reverseNode(root: RLNode?) -> RLNode {
        var newRoot: RLNode? = nil
        var rootIn: RLNode? = root
        while rootIn != nil {
            let next: RLNode? = rootIn!.next
            rootIn!.next = newRoot
            newRoot = rootIn
            rootIn = next
        }
        return newRoot!
    }
}
